### ISSUE TITLE
Mount state=mounted opts=bind mounts on every run

### ISSUE THREAD

**Issue by rbu** [ansible/ansible-modules-core#1942]
##### Ansible Configuration:

```
[ssh_connection]
pipelining = True

[defaults]
hash_behaviour = merge
```
##### Environment:

Running on Linux / Python 2.7.8. Target is CentOS 7.
##### Summary:

Running a bind mount task will mount the directory every time it is run, creating _two_ mtab entries.
##### Steps To Reproduce:

Given the following task:

```
- mount: opts=bind fstype=none state=mounted name=/dest src=/src
```

Run it twice.
##### Expected Results:
1. There is one entry in fstab:
   `/src /dest none bind 0 0`
2. There is one `mount`:
   `/dev/sda1 on /dest type btrfs (rw,relatime,ssd,space_cache)`
##### Actual Results:
1. There is one entry in fstab:
   `/src /dest none bind 0 0`
2. There are three `mount`s:

```
/dev/sda1 on /dest type btrfs (rw,relatime,ssd,space_cache)
/dev/sda1 on /dest type btrfs (rw,relatime,ssd,space_cache)
/dev/sda1 on /src type btrfs (rw,relatime,ssd,space_cache)
```
##### Notes:

This is a re-occurrence of #86. Note that [https://github.com/ansible/ansible-modules-core/blob/6a872469fc1bda54126dfa91338cf7a26bde701c/system/mount.py#L359](the special case exactly for this) that was introduced in #49 expects `mount -l` to output the source path. This does not reliably work with all distributions. Some versions/configurations of util-linux do not keep the source path in /etc/mtab, as is seen in the `mount` output above.

Note that `findmnt` lists this:

```
TARGET                                                          SOURCE             FSTYPE      OPTIONS
...
/                                                               /dev/sda1          btrfs       rw,relatime,ssd,space_cache
...
├─/dest                                                         /dev/sda1[/src]    btrfs       rw,relatime,ssd,space_cache
│ └─/dest                                                       /dev/sda1[/src]    btrfs       rw,relatime,ssd,space_cache
└─/src                                                          /dev/sda1[/src]    btrfs       rw,relatime,ssd,space_cache
```

**Comment by emielmolenaar** [comment]
Same problem here, running Ansible 1.9.2 on Debian Jessie. Targets are all Debian Jessie.

**Comment by samsun387** [comment]
having the exact same problem on Ubuntu 14.04LTS. Ansible version 1.9.4

**Comment by luto** [comment]
seeing the same on CentOS 7, ansible https://github.com/ansible/ansible/commit/8607aa527de0bceb72eff825e81ad56870a5b9d6 (2.1.0). Core Modules: 15dcabe01231d43788c05152c80eaec41bcb6dad.

**Comment by luto** [comment]
Additionally this causes the `setup` task to fail after a while, because the mtab-file expectedly becomes giant:

```
$ wc -l /etc/mtab
131101
```

**Comment by crane-denny** [comment]
Nasty workaround for it...

```
- name: Bindmounting document root into home directory
  mount:
    name: "/home/user/doc_root"
    src: "/var/www/html/site"
    opts: bind
    fstype: none
    state: present
  register: fstab

- name: Reload fstab
  command: mount -a
  when: fstab.changed
```

**Comment by piotrplenik** [comment]
Same on debian, ansible/ansible@49c15c81376c258d130cd0f89c6ff25467ebf17b

**Comment by piotrplenik** [comment]
I found, that problem is in:
https://github.com/ansible/ansible-modules-core/blob/devel/system/mount.py#L366

Line 366 is:

```
if arguments[0] == args['src'] and arguments[2] == args['name'] and arguments[4] == args['fstype']:
```

Problem is that:
- `arguments[0]` is partition name instead of src 
  -`arguments[4]` is "ext4" instead of fstype argument ("none")

My proposal is to use "findmnt -l" instead of "mount -l" command in line 361 and change checking.
I will prepare PR for this.

**Comment by piotrplenik** [comment]
This issue is related with #2924

**Comment by anisse** [comment]
I can reproduce this in ansible 2.0.10

**Comment by piotrplenik** [comment]
@anisse, have you try to use fix in 4f1336f ?

**Comment by ansibot** [comment]
@grossws, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@grossws, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by abadger** [comment]
This should be fixed in current stable-2.1 and devel for centos7 (the platform reported in this bug).  Testing this against a range of platforms, it still seems to be a problem on ubuntu12 and ubuntu14 (but not ubuntu16.04).  The error looks like we just don't have the proper information being returned by findmnt or mount there to be able to fix this.  Will continue to look for alternatives that we might be able to merge in time for 2.2 (unlikely that more fixes will make it into 2.1.x).

**Comment by abadger** [comment]
@jtyr's latest PR: https://github.com/ansible/ansible-modules-core/pull/5055 has been merged to devel.  It fixes this on the remaining platforms.  Going to close this as fixed in 2.2.

**Comment by abadger** [comment]
The previous fixes caused some additional issues.  Since we're releasing 2.1.2 later today we reverted the fixes from stable-2.1.  They will not be present in 2.1.2.  Working out the bugs in the code continues in the devel branch and will be in 2.2.0.

**Comment by killua-eu** [comment]
OK, so this bug bit me on

```
# ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
# uname -a
Linux rexe 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

```
in with this task
```
- name: bind paths
  mount: 
   opts: bind
   fstype: none
   state: mounted 
   name: "{{ vsftpd_local_root }}/{{ item.0.username }}/{{ item.1 | realpath | basename }}"
   src: "{{ item.1 | realpath }}"
  #debug: msg="Path {{ item.1 | realpath }} to {{ vsftpd_local_root }}/{{ item.0.username }}/{{ item.1 | realpath | basename }}"
  with_subelements:
    - "{{ vsftpd_users }}"
    - bindpath
    - flags:
skip_missing: true
```
... i get multiple entries with bind mount. @abadger , could you please check if you can reproduce this regression?

**Comment by jtyr** [comment]
@killua-eu Pelase could you share the content of `/proc/self/mountinfo` before and after the first bind mount?

**Comment by killua-eu** [comment]
unmounted all bind mounts in the relevant lxd container to get a clean slate. got:
```
# cat /proc/self/mountinfo    
184 98 252:0 /var/lib/lxd/containers/sites-lamp/rootfs / rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
185 184 0:44 / /dev rw,nodev,relatime - tmpfs none rw,size=492k,mode=755,uid=100000,gid=100000
186 184 0:43 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
187 184 0:45 / /sys rw,nodev,relatime - sysfs sysfs rw
188 185 0:6 /fuse /dev/fuse rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
189 185 0:6 /net/tun /dev/net/tun rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
190 186 0:46 / /proc/sys/fs/binfmt_misc rw,relatime master:78 - binfmt_misc binfmt_misc rw
191 187 0:37 / /sys/fs/fuse/connections rw,relatime master:27 - fusectl fusectl rw
192 187 0:24 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime master:11 - pstore pstore rw
198 187 0:7 / /sys/kernel/debug rw,relatime master:25 - debugfs debugfs rw
199 187 0:12 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime master:8 - securityfs securityfs rw
200 185 0:16 / /dev/mqueue rw,relatime master:26 - mqueue mqueue rw
281 185 252:0 /var/lib/lxd/devlxd /dev/lxd rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
282 185 252:0 /var/lib/lxd/shmounts/sites-lamp /dev/.lxd-mounts rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
283 186 0:40 /proc/cpuinfo /proc/cpuinfo rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
284 186 0:40 /proc/diskstats /proc/diskstats rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
285 186 0:40 /proc/meminfo /proc/meminfo rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
286 186 0:40 /proc/stat /proc/stat rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
287 186 0:40 /proc/swaps /proc/swaps rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
288 186 0:40 /proc/uptime /proc/uptime rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
289 185 0:6 /null /dev/null rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
290 185 0:6 /zero /dev/zero rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
291 185 0:6 /full /dev/full rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
292 185 0:6 /urandom /dev/urandom rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
293 185 0:6 /random /dev/random rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
294 185 0:6 /tty /dev/tty rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
295 185 0:14 /0 /dev/console rw,nosuid,noexec,relatime master:3 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
99 185 0:47 / /dev/pts rw,relatime - devpts devpts rw,gid=100005,mode=620,ptmxmode=666
100 185 0:48 / /dev/shm rw,nosuid,nodev - tmpfs tmpfs rw,uid=100000,gid=100000
101 184 0:49 / /run rw,nosuid,nodev - tmpfs tmpfs rw,mode=755,uid=100000,gid=100000
102 101 0:50 / /run/lock rw,nosuid,nodev,noexec,relatime - tmpfs tmpfs rw,size=5120k,uid=100000,gid=100000
104 187 0:51 / /sys/fs/cgroup ro,nosuid,nodev,noexec - tmpfs tmpfs ro,mode=755,uid=100000,gid=100000
105 104 0:23 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd
106 104 0:27 / /sys/fs/cgroup/cpu,cpuacct rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,cpu,cpuacct
107 104 0:34 / /sys/fs/cgroup/perf_event rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,perf_event,release_agent=/run/cgmanager/agents/cgm-release-agent.perf_event
108 104 0:26 / /sys/fs/cgroup/cpuset rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,cpuset,clone_children
109 104 0:32 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,hugetlb,release_agent=/run/cgmanager/agents/cgm-release-agent.hugetlb
110 104 0:29 / /sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,memory
111 104 0:30 / /sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,devices
112 104 0:25 / /sys/fs/cgroup/freezer rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,freezer
113 104 0:33 / /sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,blkio
114 104 0:28 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,net_cls,net_prio
115 104 0:31 / /sys/fs/cgroup/pids rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,pids,release_agent=/run/cgmanager/agents/cgm-release-agent.pids
```
after first run
```
# cat /proc/self/mountinfo
184 98 252:0 /var/lib/lxd/containers/sites-lamp/rootfs / rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
185 184 0:44 / /dev rw,nodev,relatime - tmpfs none rw,size=492k,mode=755,uid=100000,gid=100000
186 184 0:43 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
187 184 0:45 / /sys rw,nodev,relatime - sysfs sysfs rw
188 185 0:6 /fuse /dev/fuse rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
189 185 0:6 /net/tun /dev/net/tun rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
190 186 0:46 / /proc/sys/fs/binfmt_misc rw,relatime master:78 - binfmt_misc binfmt_misc rw
191 187 0:37 / /sys/fs/fuse/connections rw,relatime master:27 - fusectl fusectl rw
192 187 0:24 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime master:11 - pstore pstore rw
198 187 0:7 / /sys/kernel/debug rw,relatime master:25 - debugfs debugfs rw
199 187 0:12 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime master:8 - securityfs securityfs rw
200 185 0:16 / /dev/mqueue rw,relatime master:26 - mqueue mqueue rw
281 185 252:0 /var/lib/lxd/devlxd /dev/lxd rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
282 185 252:0 /var/lib/lxd/shmounts/sites-lamp /dev/.lxd-mounts rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
283 186 0:40 /proc/cpuinfo /proc/cpuinfo rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
284 186 0:40 /proc/diskstats /proc/diskstats rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
285 186 0:40 /proc/meminfo /proc/meminfo rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
286 186 0:40 /proc/stat /proc/stat rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
287 186 0:40 /proc/swaps /proc/swaps rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
288 186 0:40 /proc/uptime /proc/uptime rw,nosuid,nodev,relatime master:75 - fuse.lxcfs lxcfs rw,user_id=0,group_id=0,allow_other
289 185 0:6 /null /dev/null rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
290 185 0:6 /zero /dev/zero rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
291 185 0:6 /full /dev/full rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
292 185 0:6 /urandom /dev/urandom rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
293 185 0:6 /random /dev/random rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
294 185 0:6 /tty /dev/tty rw,nosuid,relatime master:2 - devtmpfs udev rw,size=488064k,nr_inodes=122016,mode=755
295 185 0:14 /0 /dev/console rw,nosuid,noexec,relatime master:3 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
99 185 0:47 / /dev/pts rw,relatime - devpts devpts rw,gid=100005,mode=620,ptmxmode=666
100 185 0:48 / /dev/shm rw,nosuid,nodev - tmpfs tmpfs rw,uid=100000,gid=100000
101 184 0:49 / /run rw,nosuid,nodev - tmpfs tmpfs rw,mode=755,uid=100000,gid=100000
102 101 0:50 / /run/lock rw,nosuid,nodev,noexec,relatime - tmpfs tmpfs rw,size=5120k,uid=100000,gid=100000
104 187 0:51 / /sys/fs/cgroup ro,nosuid,nodev,noexec - tmpfs tmpfs ro,mode=755,uid=100000,gid=100000
105 104 0:23 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd
106 104 0:27 / /sys/fs/cgroup/cpu,cpuacct rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,cpu,cpuacct
107 104 0:34 / /sys/fs/cgroup/perf_event rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,perf_event,release_agent=/run/cgmanager/agents/cgm-release-agent.perf_event
108 104 0:26 / /sys/fs/cgroup/cpuset rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,cpuset,clone_children
109 104 0:32 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,hugetlb,release_agent=/run/cgmanager/agents/cgm-release-agent.hugetlb
110 104 0:29 / /sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,memory
111 104 0:30 / /sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,devices
112 104 0:25 / /sys/fs/cgroup/freezer rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,freezer
113 104 0:33 / /sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,blkio
114 104 0:28 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,net_cls,net_prio
115 104 0:31 / /sys/fs/cgroup/pids rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,pids,release_agent=/run/cgmanager/agents/cgm-release-agent.pids
150 184 252:0 /var/lib/lxd/containers/sites-lamp/rootfs/var/www/vhosts/example_org /srv/ftp/user/example_org rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered

```

an additional run addse the example_org

```
150 184 252:0 /var/lib/lxd/containers/sites-lamp/rootfs/var/www/vhosts/example_org /srv/ftp/user/example_org rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered
296 150 252:0 /var/lib/lxd/containers/sites-lamp/rootfs/var/www/vhosts/example_org /srv/ftp/user/example_org rw,relatime master:1 - ext4 /dev/mapper/vg-lv_root rw,errors=remount-ro,data=ordered

```

**Comment by tace** [comment]
I'm using ansible 2.3.2.0 and still hitting this issue (RedHat), so is this fixed or not as this ticket is closed?

**Comment by willthames** [comment]
According to this issue, it was thought to be fixed. If you are still experiencing problems, open a new issue on the ansible/ansible repo (as all open issues have been migrated from here to there now anyway)

**Comment by tace** [comment]
@willthames thanks for your info. Found it there, so it's still open: https://github.com/ansible/ansible/issues/29870
